In [1]:
import pandas as pd
import numpy as np
from pandas import read_csv

In [2]:
df=read_csv("/content/Crop_recommendation.csv")

In [3]:
df.head()

,N,P,K,temperature,humidity,ph,rainfall,label
0,90,42,43,20.879744,82.002744,6.502985,202.935536,rice
1,85,58,41,21.770462,80.319644,7.038096,226.655537,rice
2,60,55,44,23.004459,82.320763,7.840207,263.964248,rice
3,74,35,40,26.491096,80.158363,6.980401,242.864034,rice
4,78,42,42,20.130175,81.604873,7.628473,262.717340,rice


In [5]:
# Get the mapping from the LabelEncoder
label_mapping = dict(zip(le.classes_, le.transform(le.classes_)))

# Display the mapping
print("Mapping of encoded labels to actual labels:")
print(label_mapping)

Mapping of encoded labels to actual labels:
{'apple': np.int64(0), 'banana': np.int64(1), 'blackgram': np.int64(2), 'chickpea': np.int64(3), 'coconut': np.int64(4), 'coffee': np.int64(5), 'cotton': np.int64(6), 'grapes': np.int64(7), 'jute': np.int64(8), 'kidneybeans': np.int64(9), 'lentil': np.int64(10), 'maize': np.int64(11), 'mango': np.int64(12), 'mothbeans': np.int64(13), 'mungbean': np.int64(14), 'muskmelon': np.int64(15), 'orange': np.int64(16), 'papaya': np.int64(17), 'pigeonpeas': np.int64(18), 'pomegranate': np.int64(19), 'rice': np.int64(20), 'watermelon': np.int64(21)}


In [6]:
from sklearn.preprocessing import LabelEncoder

le = LabelEncoder()
df['label_encoded'] = le.fit_transform(df['label'])

display(df.head())

,N,P,K,temperature,humidity,ph,rainfall,label,label_encoded
0,90,42,43,20.879744,82.002744,6.502985,202.935536,rice,20
1,85,58,41,21.770462,80.319644,7.038096,226.655537,rice,20
2,60,55,44,23.004459,82.320763,7.840207,263.964248,rice,20
3,74,35,40,26.491096,80.158363,6.980401,242.864034,rice,20
4,78,42,42,20.130175,81.604873,7.628473,262.717340,rice,20


## Manual Testing

### Subtask:
Test the trained model with new input data and interpret the prediction.

**Reasoning**:
Create a sample input with the same features as the training data, make a prediction using the trained model, and then use the label mapping to convert the numerical prediction back to the original label.

# Task
Train a classification model on the dataset.

## Split data

### Subtask:
Split the dataframe into features (X) and the target variable (y). Then split these into training and testing sets.


**Reasoning**:
Split the dataframe into features (X) and the target variable (y), then split these into training and testing sets.



In [7]:
from sklearn.model_selection import train_test_split

X = df.drop(['label', 'label_encoded'], axis=1)
y = df['label_encoded']

X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42, stratify=y)

## Import and initialize model

### Subtask:
Import a classification model (e.g., RandomForestClassifier) and initialize it.


**Reasoning**:
Import the RandomForestClassifier and instantiate it.



In [8]:
from sklearn.ensemble import RandomForestClassifier

model = RandomForestClassifier(n_estimators=100, random_state=42)

## Train the model

### Subtask:
Train the initialized model using the training data.


**Reasoning**:
Train the initialized model using the training data.



In [9]:
model.fit(X_train, y_train)

RandomForestClassifier(random_state=42)

## Evaluate the model

### Subtask:
Make predictions on the test set and evaluate the model's performance using appropriate metrics.


**Reasoning**:
Make predictions on the test set and evaluate the model's performance using appropriate metrics.



In [10]:
from sklearn.metrics import accuracy_score, precision_score, recall_score, f1_score

y_pred = model.predict(X_test)

accuracy = accuracy_score(y_test, y_pred)
precision = precision_score(y_test, y_pred, average='weighted')
recall = recall_score(y_test, y_pred, average='weighted')
f1 = f1_score(y_test, y_pred, average='weighted')

print(f"Accuracy: {accuracy}")
print(f"Precision: {precision}")
print(f"Recall: {recall}")
print(f"F1-score: {f1}")

Accuracy: 0.9954545454545455
Precision: 0.9956709956709956
Recall: 0.9954545454545455
F1-score: 0.9954517027687759


## Summary:

### Data Analysis Key Findings

* The data was split into training and testing sets, with 20% of the data allocated for testing.
* A RandomForestClassifier model was initialized with 100 estimators and a random state of 42.
* The model was trained on the training data (\(X\_train\) and \(y\_train\)).
* The model achieved high performance on the test set, with an accuracy of 0.9955, precision of 0.9957, recall of 0.9955, and an F1-score of 0.9955.

### Insights or Next Steps

* The model shows excellent performance on the test data, suggesting it generalizes well to unseen data.
* Further investigation could involve exploring the most important features identified by the RandomForestClassifier to understand which features contribute most to the predictions.


In [12]:
# Example new data for prediction (replace with your desired values)
new_data = [[90, 42, 43, 20.879744, 82.002744, 6.502985, 202.935536]]

# Convert the new data to a pandas DataFrame
new_data_df = pd.DataFrame(new_data, columns=X_train.columns)

# Make a prediction
predicted_label_encoded = model.predict(new_data_df)

# Convert the encoded label back to the original label
predicted_label = [key for key, value in label_mapping.items() if value == predicted_label_encoded[0]][0]

print(f"The predicted crop is: {predicted_label}")

The predicted crop is: rice


## Interactive Prediction

### Subtask:
Create interactive input fields for the user to enter crop characteristics and get a prediction.

**Reasoning**:
Use `ipywidgets` to create input fields for each feature, collect the user's input, make a prediction using the trained model, and display the predicted crop.

In [13]:
from ipywidgets import FloatText, Button, VBox, Output
from IPython.display import display

# Create input widgets for each feature
n_input = FloatText(description='Nitrogen (N):')
p_input = FloatText(description='Phosphorus (P):')
k_input = FloatText(description='Potassium (K):')
temp_input = FloatText(description='Temperature:')
humidity_input = FloatText(description='Humidity:')
ph_input = FloatText(description='pH:')
rainfall_input = FloatText(description='Rainfall:')

# Create a button to trigger the prediction
predict_button = Button(description='Predict Crop')

# Create an output widget to display the result
output = Output()

# Function to handle button click
def on_predict_button_clicked(b):
    with output:
        output.clear_output()
        try:
            # Get input values
            input_data = [[n_input.value, p_input.value, k_input.value,
                           temp_input.value, humidity_input.value, ph_input.value,
                           rainfall_input.value]]

            # Convert to DataFrame
            input_df = pd.DataFrame(input_data, columns=X_train.columns)

            # Make prediction
            predicted_label_encoded = model.predict(input_df)

            # Convert encoded label back to original label
            predicted_label = [key for key, value in label_mapping.items() if value == predicted_label_encoded[0]][0]

            print(f"The predicted crop is: {predicted_label}")
        except Exception as e:
            print(f"An error occurred: {e}")

# Link the button click to the prediction function
predict_button.on_click(on_predict_button_clicked)

# Arrange widgets vertically
input_widgets = VBox([n_input, p_input, k_input, temp_input, humidity_input, ph_input, rainfall_input, predict_button, output])

# Display the widgets
display(input_widgets)